lib/JMAP/Tester.pm

```diff
@@ -78,6 +79,19 @@ There is also L<JMAP::Tester::Response/"as_stripped_pairs">.
 
 =cut
 
+has return_futures => (
+  is  => 'ro',
+  default => 0,
+);
+
+my $Failsafe = sub {
```

Comment:
Can we get some comments here to explain the purpose?

Reply:
Yes!

When something doesn't work — not an individual method call, but the whole HTTP call, somehow — then the future will fail, and the failure might be a JMAP tester failure object, meaning we semi-expected it, or it might be some other crazy failure, meaning we had no way of seeing it coming.

We use Future->fail because that way we can use ->else in chains to only act on successful HTTP calls.  At the end, it's fine if you're expecting a future and can know that a failed future is a fail and a done future is okay.  In the old calling convention, though, you expect to get a success/fail object no matter what, and unexpected behavior is fatal.

`$Failsafe` emulates that.  Futures being resolved before return are converted like so:

* successful futures return their payload, the Result object
* failed futures that contain a JMAP tester Failure return the failure
* other failed futures die, like they would if you called `$failed_future->get`

Reply:
Thanks for the explanation! Can we get some condensed form of that as code comments in JMAP::Tester? (Perhaps my original comment wasn't entirely clear there!) :)

Reply:
I've pushed up an explanation.